### ISSUE TITLE
Add a simulation layer for MoveOS?

### ISSUE THREAD

**Issue by wubuku** [rooch-network/rooch#100]
Consider a scenario in a consortium blockchain where multiple validator nodes exist. When a transaction involves shared objects, it is clear that "consensus needs to be reached".

In terms of performance optimization, more or less, there are always some times when transactions need to go through the consensus process, which is an inherent disadvantage that web3 applications cannot avoid compared to web2 applications.

If a layer can be built on MoveOS to perform instant simulation (dry run) of transactions, it could be very helpful for this issue.

Specifically, after a transaction is submitted to a node, it will be immediately simulated and the resulting consequences (including events generated by the simulation and modifications to the state) can be queried through APIs provided by the node.

This goes beyond the simulate(dry run)-transaction API that some Move-based blockchains have already provided and is widely used by wallets.

Developers can consider immediately querying the "possible state" of the system after submitting a transaction to the node, present it to the user in an appropriate way, and then guide the user to "start building the next transaction based on this simulated execution result".

What are the benefits of providing such a simulation layer to developers? This question is perhaps worth discussing.

Let me try to mention a few.

For example, as a user of an e-commerce DApp, if you want to recharge an account and then purchase some goods, after sending the transfer-coin transaction, you may not need to wait for the recharge to arrive before building the "place an order" transaction.

Although this approach does not accelerate consensus achievement, it is evident that at the very least, it improves the user experience. User experience is the most important aspect.

Of course, in this scenario, the code implementation does not necessarily need the support of a simulation layer. However, with the support of a simulation layer, it may be easier to implement.

The challenge in implementation may lie in how to design and develop a system that can simulate transaction execution rapidly and accurately (as much as possible, given the inevitable differences between the simulated and actual execution results). In addition, when a submitted transaction "actually fails" during actual execution, it is necessary to roll back the state of the simulation layer to a "suitable state".

**Comment by wubuku** [comment]
Perhaps the example I gave above was too simple to fully showcase the value of this feature. Need thinking of more scenarios where it can be used.

Overall, with this simulation layer, developers can develop a type of functionality as follows:

A complex business process that requires users to perform multiple steps and submit multiple transactions. The conventional approach is to make users wait for consensus to be reached and transactions to be completed after each transaction submission, before allowing users to proceed to the next step. This may harm the user experience. 

With the simulation layer, we can use two methods to enhance the user experience based on specific needs: 

1. We can allow the user to proceed to the next step (building the next transaction) without waiting for the previous transaction to complete by assuming that the transaction will be executed successfully. 
2. We can first simulate the entire process with the user (of course, the user needs to be explicitly notified that this is a "simulation" state), then present the simulation results to the user for confirmation. If the user confirms the execution, we can batch-submit the constructed multiple transactions.

**Comment by jolestar** [comment]
Rooch can execute and return results immediately, do not need to wait for the consensus

**Comment by wubuku** [comment]
![image](https://github.com/rooch-network/rooch/assets/4245538/83fb4d36-526e-42c7-b7f1-6f7ac51a8f63)

Based on my understanding of Rooch, from a micro perspective, only by submitting the sequenced transactions to MoveOS for execution can you immediately get the results - that is, you can 'Get' the current state after execution. (And according to my understanding, MoveOS primarily refers to the core of Proposer. Right?)

Indeed, by embedding MoveOS as the execution environment for the Move language within a centralized application, we can fully enjoy the benefits brought by the Move language. Moreover, the system's performance should not be adversely affected since it is centralized.

That's true. If we are building a decentralized application using Rooch, we may need more than just MoveOS. We may also need to use **other modules** or applications provided by Rooch, or even customize our own modules to fit our specific needs. For example, Sequencer is essential in creating a decentralized application, and the process of confirming the final state may become more complex.

As application developers, they should expect to work **on top of the APIs provided by the "whole"** comprised of these Rooch modules, rather than just directly facing MoveOS. The application level code expects to submit a transaction to the API of "the whole" and get the final state as quickly as possible.

The higher the level of decentralization required for our application, the greater the "consensus cost" we need to pay. This seems inevitable. The cost also depends on the number of shared objects needed for the application's transactions and the cost required to maintain these shared objects.

For example, a time-sensitive decentralized application may need to maintain a decentralized `clock` object with sufficient accuracy to meet the application's requirements. Based on my understanding of Rooch, this could result in a large number of transactions, which would still need to be processed by a sequencer. This will inevitably have perceptible effects on the performance of executing other transactions. Right?

In summary, the simulation layer I mentioned refers to something that is located below the API provided by "the whole" but runs on top of the Move execution engine. It hides the complexity of how various modules in Rooch work together from developers and can provide "continuous dry run results".

**Comment by jolestar** [comment]
The DApp developer does not need to understand the sequencer and proposer. He can treat rooch as the simulation layer which worked as you expected.

We need to write a document describing the transaction flow, and the transaction confirmation time.

**Comment by jolestar** [comment]
We do not need this. Just need to provide a transaction simulation API #422.
